### ISSUE TITLE
overshadowImplementation with JsonDeserialize and static builder method fails

### ISSUE THREAD

**Issue by cole-snodgrass** [immutables/immutables#645]
Not quite the same as issue #468, but maybe semi related? The following interface fails to generate valid java code:
```
@Value.Immutable(copy = false)
@Value.Style(
    stagedBuilder = true,
    optionalAcceptNullable = true,
    visibility = Value.Style.ImplementationVisibility.PACKAGE,
    overshadowImplementation = true
)
@JsonDeserialize(as = ImmutableExample.class)
public interface Example {
    String name();

    class Builder extends ImmutableExample.Builder {}
    static Builder builder() {
        return new Builder();
    }
}
```

Compilation will fail with the error
```
ImmutableExample.java:106: error: incompatible types: ImmutableExample.Builder cannot be converted to Example.Builder
```

And the offending line of code is
```
103  @Deprecated
104  @JsonCreator(mode = JsonCreator.Mode.DELEGATING)
105  static ImmutableExample fromJson(Json json) {
106    Example.Builder builder = ((ImmutableExample.Builder) new Example.Builder());
107    if (json.name != null) {
108      builder.name(json.name);
109    }
110    return (ImmutableExample) builder.build();
111  }
```

I _think_ in this case the line should instead be `Example.Builder builder = new Example.Builder();`, but I haven't verified this.

This is with Immutables version `2.5.3`.

**Comment by cole-snodgrass** [comment]
This actually appears to be an issue with the `stagedBuilder`. Instead if a `strictBuilder` is used, the compilation error goes away.

**Comment by elucash** [comment]
thank you for reporting this! I fixed wrong code generated, but also figured out that staged builder and extending builder are not to be used together, thus there's a lengthy warning message added, which read like this:
```
/src/org/immutables/fixture/style/ExtendingAndStagedBuilder.java:30: warning:
Extending Builder shouldn't be used with stagedBuilder style attribute, they are
incompartible: Staged builder generate series of staged interfaces, but extending
builder actually extends implementation and do not provide type safety for setting
first attribute, as well as stagedBuilder forces generated builder interfaces to leak
in code using the builder and hence defeating the purpose of using extending builder.
  class Builder extends ImmutableExtendingAndStagedBuilder.Builder {}
  ^
```
to be released in 2.5.5

**Comment by elucash** [comment]
released in 2.5.5 thank you!
